lib/util/mp4_generator.js

```diff
@@ -0,0 +1,988 @@
+    ]);
+    for (const drmInfo of this.stream_.drmInfos) {
+      if (drmInfo && drmInfo.keyIds && drmInfo.keyIds.size) {
+        for (const keyId of drmInfo.keyIds) {
+          defaultKeyId = this.hexStringToBuffer_(keyId);
+        }
+      }
+    }
+
+    const Mp4Generator = shaka.util.Mp4Generator;
+    const bytes = new Uint8Array([
+      0x00, // version 0
+      0x00, 0x00, 0x00, // flags
+      0x00, 0x00, // Reserved fields
+      0x01, // Default protected: true
+      0x08, // Default per-sample IV size: 8
+    ]);
+    goog.asserts.assert(defaultKeyId, 'Default KID should be non-null');
+    return Mp4Generator.box('tenc', bytes, defaultKeyId);
+  }
+
+  /**
+   * Convert a hex string to buffer.
+   *
+   * @param {string} str
+   * @return {Uint8Array}
+   * @private
+   */
+  hexStringToBuffer_(str) {
+    const buf = new Uint8Array(str.length / 2);
+    for (let i = 0; i < str.length / 2; i += 1) {
+      buf[i] = parseInt(String(str[i * 2] + str[i * 2 + 1]), 16);
+    }
+    return buf;
+  }
+
+  /**
+   * Convert a string to char code.
+   *
+   * @param {string} str
+   * @return {number}
+   * @private
+   */
+  stringToCharCode_(str) {
+    let code = 0;
+    for (let i = 0; i < str.length; i += 1) {
+      code |= str.charCodeAt(i) << ((str.length - i - 1) * 8);
+    }
+    return code;
+  }
+
+  /**
+   * @private
+   */
+  static initStaticProperties_() {
+    const Mp4Generator = shaka.util.Mp4Generator;
+    if (Mp4Generator.initializated_) {
+      return;
+    }
+
+    Mp4Generator.initializated_ = true;
+
+    const majorBrand = new Uint8Array([105, 115, 111, 109]); // isom
+    const avc1Brand = new Uint8Array([97, 118, 99, 49]); // avc1
+    const minorVersion = new Uint8Array([0, 0, 0, 1]);
+
+    Mp4Generator.FTYP_ = Mp4Generator.box(
+        'ftyp', majorBrand, minorVersion, majorBrand, avc1Brand);
+    const drefBox = Mp4Generator.box('dref', Mp4Generator.DREF_);
+    Mp4Generator.DINF_ = Mp4Generator.box('dinf', drefBox);
+  }
+
+  /**
+   * Generate a box
+   *
+   * @param {string} boxName
+   * @param {...!Uint8Array} payload
+   * @return {!Uint8Array}
+   */
+  static box(boxName, ...payload) {
+    let type = shaka.util.Mp4Generator.BOX_TYPES_[boxName];
+    if (!type) {
+      type = [
+        boxName.charCodeAt(0),
+        boxName.charCodeAt(1),
+        boxName.charCodeAt(2),
+        boxName.charCodeAt(3),
+      ];
+      shaka.util.Mp4Generator.BOX_TYPES_[boxName] = type;
+    }
+    let size = 8;
+    let i = payload.length;
+    const len = i;
+    // calculate the total size we need to allocate
+    while (i--) {
+      size += payload[i].byteLength;
+    }
+
+    const result = new Uint8Array(size);
+    result[0] = (size >> 24) & 0xff;
+    result[1] = (size >> 16) & 0xff;
+    result[2] = (size >> 8) & 0xff;
+    result[3] = size & 0xff;
+    result.set(type, 4);
+    // copy the payload into the result
+    for (i = 0, size = 8; i < len; i++) {
```

Comment:
Re-using the local variable `size` for a completely different purpose here seems confusing.
Why not just call this `pointer` or something?

Reply:
Renamed to `pointer`

Reply:
Done!